File: lab3/FileReverse.c
/*
*  Austin Yen
*  4/26/15
*  FileReverse.c
*  Reads input file, reverses the order of letters in each word, and 
*  prints each word on a separate line of the output file.
*/
#include<stdio.h>
#include<stdlib.h>

/*
*  Reverses order of letters in a given string
*/
void stringReverse(char* s){
	int i, j;
	i = 0;
	j = strlen(s)-1;
	while(i < j){
		char temp = s[j];
		s[j] = s[i];
		s[i] = temp;
		i++;
		j--;
	}
}

int main(int argc, char* argv[]){
	FILE* in; /* file handle for input */ 
	FILE* out; /* file handle for output*/
	char word[256]; /* char array to store words from input file */

	/* check command line for correct number of arguments */

	if( argc != 3 ){
		printf("Usage: %s <input file> <output file>\n", argv[0]);
		exit(EXIT_FAILURE);
	}
	/* open input file for reading */
	in = fopen(argv[1], "r");
	if( in==NULL ){
		printf("Unable to read from file %s\n", argv[1]);
		exit(EXIT_FAILURE);
	}
	/* open output file for writing */
	out = fopen(argv[2], "w");
	if( out==NULL ){
		printf("Unable to write to file %s\n", argv[2]);
		exit(EXIT_FAILURE);
	}
	/* read words from input file, print on separate lines to output file*/
	while(fscanf(in, " %s", word) != EOF){
		stringReverse(word);
		fprintf(out, "%s\n", word);
	}
	/* close input and output files */
	fclose(in);
	fclose(out);
	return(EXIT_SUCCESS);
}

File: lab2/FileReverse.java
// Austin Yen
// 4/17/15
// FileReverse.java
// Given an input file, returns an output file 
// with string contents of input file reversed
import java.io.*;
import java.util.Scanner;
public class FileReverse {
	public static void main(String[] args)throws IOException{
		Scanner in = null;
		PrintWriter out = null;
		String line = null;
		String[] token = null;
		int i, n, lineNumber = 0;
		// check command line arguments
		if(args.length < 2){
			System.out.println("Usage: FileTokens infile outfile");
			System.exit(1);
		}
		// open files
		in = new Scanner(new File(args[0]));
		out = new PrintWriter(new FileWriter(args[1]));
		// read lines from in, extract and print tokens from each line
		while( in.hasNextLine() ){
			lineNumber++;
			// trim leading and trailing spaces, then add one trailing space so 
			// split works on blank lines
			line = in.nextLine().trim() + " "; 
			// split line around white space 
			token = line.split("\\s+");  
			// prints each reversed token on a new line in output file
			for (int j = 0; j < token.length; j++){
				out.println(stringReverse(token[j],token[j].length()));
			}	
		}
		
		// close both files
		in.close();
		out.close();
	}
	
	// recursively reverses the first n characters of s
	public static String stringReverse(String s, int n){
		if (n > 0){
			return (stringReverse(s.substring(1), n-1) + s.charAt(0));
		}
		return "";
	}
}
